### ISSUE TITLE
Port Widescreen from Russian Doom

### ISSUE THREAD

**Issue by drfrag666** [drfrag666/RUDE#2]
@fabiangreffrath @JNechaevsky
I want to port Widescreen from Rusian Doom but i want to make it an option and to work only with max screen blocks, the idea is to keep it as simple as possible. I also want to add the mugshot to that mode to make it more extreme but keep some info on your health level.
I'm afraid i'll need some help so i'll post questions here. Regards.

**Comment by JNechaevsky** [comment]
Sadly, I was never lucky in consolidating wide/non-wide modes into one executable, that's why I'm making two separate executables: https://github.com/JNechaevsky/russian-doom/blob/master/src/i_video.h#L31 (uncommenting `#define WIDESCREEN` before making a widescreen build).

**Comment by drfrag666** [comment]
I know but i want to make one exe so we'll see how it goes. Also no mouselook or highres here.

**Comment by fabiangreffrath** [comment]
You might want to get rid of this bug first:  https://github.com/JNechaevsky/russian-doom/issues/154

**Comment by drfrag666** [comment]
@bradharding
Ouch! I hadn't noticed. I see Doom Retro is not affected but it's 16:10 and the code was in from the first commit.

**Comment by drfrag666** [comment]
I haven't reached very far. It crashes in R_DrawSpan, seems it requires MBF code. I've got 32 bit integer math but i keep the original code. :(
```
#0  0x0042b21c in R_DrawSpan () at C:\DEV\rude\src\doom\r_draw.c:634
#1  0x00453303 in R_MapPlane (y=-1, x1=104, x2=104) at C:\DEV\rude\src\doom\r_plane.c:171
#2  0x00453751 in R_MakeSpans (x=105, t1=4294967295, b1=4294967295, t2=4294967295, b2=3131961357) at C:\DEV\rude\src\doom\r_plane.c:365
#3  0x00453aaa in R_DrawPlanes () at C:\DEV\rude\src\doom\r_plane.c:465
#4  0x0043de15 in R_RenderPlayerView (player=0x534100 <players>) at C:\DEV\rude\src\doom\r_main.c:918
#5  0x00425c2c in D_Display () at C:\DEV\rude\src\doom\d_main.c:257
#6  0x00426137 in D_RunFrame () at C:\DEV\rude\src\doom\d_main.c:459
#7  0x00426223 in D_DoomLoop () at C:\DEV\rude\src\doom\d_main.c:511
#8  0x00427ec6 in D_DoomMain () at C:\DEV\rude\src\doom\d_main.c:2047
#9  0x004013c1 in SDL_main (argc=argc@entry=1, argv=argv@entry=0x30000) at C:\DEV\rude\src\i_main.c:67
#10 0x0046813a in main_getcmdline () at ../src/main/windows/SDL_windows_main.c:175
#11 0x00468235 in WinMain@16 (hInst=0x400000, hPrev=0x0, szCmdLine=0xd73b64 "", sw=10) at ../src/main/windows/SDL_windows_main.c:204
#12 0x00468ebb in main ()
 
```

**Comment by drfrag666** [comment]
About that bug this in r_main looks suspicious:
```
#ifdef WIDESCREEN
    tempCentery = viewheight/2 + (pitch << (hires && !detailshift)) * (screenblocks < 9 ? screenblocks : 9) / 10;
#else
    tempCentery = viewheight/2 + (pitch << (hires && !detailshift)) * (screenblocks < 11 ? screenblocks : 11) / 10;
#endif
```

**Comment by JNechaevsky** [comment]
@drfrag666 nope, MBF code is not necessary. First of all, try to set screen size to full screen mode (i.e. w/o status bar) before starting the game. Couple of things must be done before increasing `SCREENWIDTH` macro.

Also, my widescreen mode does not have a "bordered" view modes, and status bar is drawn a bit different: even when it's active, effective screen size is still same to full "screen mode".

The code you have quoted is necessary for proper slope alignment while active mouse look, AFAIR.

**Comment by drfrag666** [comment]
I've already done that. I've disabled the range check in r_draw and changed scaledviewwidth to 426 in r_main. I've reviewed "Merge wide screen code into main brach".
 About one exe instead of redefining ORIGWIDTH you could define a new one and use regular conditionals with the option where required plus disabling the appropiate rangle checks with && !defined WIDESCREEN.

**Comment by drfrag666** [comment]
I've changed the macro but i still get the same crash. With Crispy it runs and i get the correct view angle, this is just a quick lowres test. So i still think it requires MBF code or enhance the original code myself which obviously i can't do. Should i contact Lee Killough to reinvent the wheel? :)

**Comment by drfrag666** [comment]
Oops, now it works. I've set width to 420 instead to 426 by pure luck and now it doesn't crash. The old code must be less tolerant. So back on track i guess?

**Comment by drfrag666** [comment]
@fabiangreffrath @JNechaevsky
Ugh, this is quite complicated. What i wrote ealier didn't make much sense. I'm using an actualwidth static variable in i_video.cpp to replace SCREENWIDTH setting it in I_InitGraphics but now even the title screen is screwed.  It doesn't matter if i initialize it to 0 or 1. What could be wrong?

**Comment by drfrag666** [comment]
Finally i managed to get it working but it's a bit rough. I used the widescreen global and a couple of static variables in i_video (actualwidth) and v_video (screenwidth), then either locals or conditionals everywhere else. I've tried to keep it as less intrusive as possible so everything is still at the left part of the screen, and no black background. And finally i set width to 400 so it works better on 16:10 and you don't lose too much vertical view. It's only for Doom. I haven't tested much but seems it's working well and demos play.
@Julia Hey you fooled me at first with your reduced size weapons but then i realized it was a zoomed view and not real widescreen.

**Comment by JNechaevsky** [comment]
> Hey you fooled me at first with your reduced size weapons but then i realized it was a zoomed view and not real widescreen.

Wha-? Where? When?

**Comment by drfrag666** [comment]
I'm joking obviously. :D Weapons are smaller in Russian Doom.

**Comment by JNechaevsky** [comment]
Smaller than where?

**Comment by drfrag666** [comment]
Smaller than in my version and Retro. For me everything is zoomed, in your version they are not.

**Comment by drfrag666** [comment]
@bradharding
Hey, i've been looking at Doom Retro's implementation and looks pretty clean but i wonder what's the magic behind I_ToggleWidescreen. To get the correct aspect ratio the logical resolution must be 320x240 but with SDL_RenderSetLogicalSize you set 320x200 for 16:10 so the game would look stretched horizontally. Where do you re-scale the view to get the correct aspect ratio?

**Comment by bradharding** [comment]
Hi @drfrag666. DR's widescreen implementation is a bit of a hack, really. In I_ToggleWidescreen() you'll note `src_rect.h` is also changed after the call to `SDL_RenderSetLogicalSize()`, which compensates for the stretching.

**Comment by drfrag666** [comment]
Thanks. Hack or not it's a smart trick anyway, it's much cleaner than changing the screen width everywhere and you don't need to restart. So i guess the resize is handled in I_WindowResizeBlit.
Too bad that low level code is not present in Choco.

**Comment by drfrag666** [comment]
@fabiangreffrath @JNechaevsky
Well, now i'd like to solve the mystery of the scaled down player weapons. I thought it was some nice Julia's trick but now i'm not sure and may be it's due to some Crispy enhancement. Here and in Doom Retro everything is zoomed in (as expected) but in Russian Doom the weapon has the same scale on both wide and non wide modes and it looks better. I've been reviewing R_DrawPSprite but no idea of what's going on.

**Comment by JNechaevsky** [comment]
@drfrag666 
My `R_DrawPSprite` is similar to original one, except additions from Crispy and special conditions for Jaguar game mode.  Nothing extra ordinary.

**Comment by drfrag666** [comment]
Thanks, i reviewed it and Crispy's too. Must be something related to hires support, seems weapon scaling is independent of screen width somehow. No big deal anyway.
BTW Fabian is implementing is own version too.

**Comment by drfrag666** [comment]
Actually it's a problem but it's not the size of the weapon. The scaling quality is very poor and it's specially noticeable on the weapon, it looks ugly and shitty. I guess that's becouse Choco scales the image twice. Doom retro uses a different technique for wide screen and looks good. I've changed width to 400 and looks like shit even 384 does.

**Comment by drfrag666** [comment]
Oops fixed, the weapon scale is also adjusted in r_main and i missed it. I've also adjusted the light levels. Now it looks okay but it's like you're a dwarf in a world of trolls. It's better only due to the poor scaling.

**Comment by JNechaevsky** [comment]
What exactly happening? Provide a screenshot please.

**Comment by drfrag666** [comment]
It's a bit complicated. It's okay but since we increased the screen width without increasing the FOV everything (the world) is zoomed in. Turns out this is a problem with the weapon scaling code itself and not the SDL or Choco. The right scale is the one relative to SCREENWIDTH and not WIDESCREENWIDTH so the weapon is bigger as in Doom Retro but the only scale that looks good is the one so the weapon size relative to the full screen is the same as before and then the doomguy looks like a dwarf. But hey it's not too bad. :D 
![DOOM00](https://user-images.githubusercontent.com/26112290/73378017-41019c00-42c0-11ea-8329-25c05f2e95f9.jpg)

**Comment by drfrag666** [comment]
@fabiangreffrath @JNechaevsky
I see fabian has implemented real widescreen, he has beat us to it. Congrats!
I can't see anything in Crispy, now i get this but i've fixed the sliding flats problem, it was the same problem as with projection and sprites. Hope this helps. Check my master branch.
![DOOM0000](https://user-images.githubusercontent.com/26112290/73591647-8062ff00-44f1-11ea-8d42-976e9776f06a.png)

**Comment by fabiangreffrath** [comment]
Did you change aspect ratio correction in-game? I'll have to take this out of the menu, it doesn't make any sense anymore.

**Comment by drfrag666** [comment]
No, it was enabled in the ini. Changing it makes no difference, still screwed but PCX screenshots are correct.
 My version is simpler and i use a 400x240 screen, it works both with and without correction. I think the ratio is right.

**Comment by fabiangreffrath** [comment]
Ah, this also happens if you switch hires rendering while in widescreen mode. Somehow the screen pitch is still calculated wrong in this case.

**Comment by drfrag666** [comment]
I did it (both zoomed view and real wide) and seems it's working well.
